### ISSUE TITLE
[Placeholder] Refactor

### ISSUE THREAD

**Issue by BerkieBb** [qbcore-framework/qb-vehicleshop#143]
I'm going to do a bit of a refactor on this to fix a lot of the current issues, make the configuration a bit easier if that's possible and optimize it.

**Comment by Brusein** [comment]
<3

**Comment by adweex** [comment]
**Server Artifact: 5181
Server Build: 2545**
Our test server is always up to date with the latest QB resources. We are manually updateing everyday.
Changes: Added multiple vehicleshops as free-use, and there is no "managed" vehicleshops.
Finances commented out in the menu options
Test Drive commented out in the menu options
Config attached as txt

**[	Test Drive Polyzone Issue	]**
Whenever you finished with Test Drive you get teleported back where you first pushed the Test Drive button in the menu. The moment you get ported back polyzone / vehicleshop will break immediately on client side (for us without errors) Menu prompts will not work until relog. For some the script did not broke, but they had to leave the poly area and re enter it for the functions to work

**[	Swap Issue	]**
On our server there are multiple vehicleshops created:
- Basic GTA cars
- Addon cars without tuning options
- Addon cars with tuning options
- Planes & Helicopters
- Emergency Vehicles POLICE
- Emergency Vehicles MEDIC
- Emergency Vehicles MECHANIC
- Truck vehicles
If two person is staying in two different vehicleshop (or more) and one is swapping a car to another type of car, the changes happen in both Client. Try to imagine if someone swaps a plane, and it appears in the Basic GTA vehicleshop - interestingly planes are not in the Basic GTA cars. buntyverma69 fixed this issue, tested it and it works.
https://github.com/qbcore-framework/qb-vehicleshop/pull/113

**[	getVehBrand() function issue	]**
Error in line client.lua line 107
The moment a person walks into a polyzone the error appears. Interestingly sometimes cars appear, sometimes not appear in the showroom, but the error remains the same and breaks the vehicleshop
Some reported that it happens with Addon Cars - On our server addon cars work completely fine.
However Planes and Helicopters produce the same issue on our server, so we had to remove it.
During the tests Polyzone Debug was enabled. The creation of Polyzone and the Vehicleshop Succeeded, but retrieving the cars brand did not!
No idea what is the cause of this bug.

**[	Brand DeSync?		]**
Sometimes Brands desync. The moment you swap your car to another car, Menu does not update the brand, resulting in a desync in the shown car, and the brand. Purchasing a car will result with an unwanted car.
Example: You want to buy a Blista, so you swap from the t20 to the Blista. The car will be swapped, but the menu header still shows the t20. If you click on purchase the t20 will be purchased not the Blista. Only relogging fixes this issue right now.

**[	Latest VehicleShop Update	]**
This update was like 8 days ago. I tried it on a test server and broke the whole vehicleshop for us so it is not yet Implemented into the Live Server. Addon cars not working anymore :/

--- Hope it helps, if you have any questions feel free to ask. I'll try to test the whole script as much as possible and give much more detailed descriptions about these issues... and sorry for grammar issues... :'D i'll try to fix that as well.
[config.txt](https://github.com/qbcore-framework/qb-vehicleshop/files/8253324/config.txt)

**Comment by erikmeyer08** [comment]
Looking forward to the rework of the script and resolving these issue! Thank you for your time and support!

**Comment by adweex** [comment]
[Car Purchase and Despawn] 

Upon buying a new vehicle, the car is saved in the sql with a value of NULL. If somehow Fivem despawns the vehicle or server is restarted before the player stores it in a garage, the Vehicle is lost forever without a developer intervention. Default Garage needs to be set upon purchase.

**Comment by adweex** [comment]
Well, not long ago I updated to the latest version. Most of the bug just went away. Car despawn was solved by an SQL update. Test drive still broken for us.

Anyway i think i mentioned a bug where once the car spawns, the menu header may not reflect the true model / brand and this causes that the purchased vehicle will be the one that the header says not what you see in the showroom.

Solution: This bug happens with weaker PCs. Client checks if the car model has loaded. Once loaded it proceeds forward. I made a while loop after spawning the car. The menu only pops up once the car has loaded and spawned as well. This fixed the issue and now everything seems functioning well. Ofc players with weaker machines sometimes have to wait 3-5 second before any menu shows up.
